test/integration/targets/mongodb_shard/tasks/main.yml

```diff
@@ -0,0 +1,367 @@
+     - "localhost:3006"
+
+- name: Ensure /home/tests/config dir exists
+  file:
+    path: /home/tests/config
+    state: directory
+
+- name: Launch cfg server
+  command: mongod --configsvr --port 4000 --dbpath /home/tests/config --logpath /home/tests/config.log --smallfiles --replSet "{{ configsrv_replicaset }}" --fork
+
+- name: Create config srv replicaset with module
+  mongodb_replicaset:
+    login_user: "{{ mongodb_admin_user }}"
+    login_password: "{{ mongodb_admin_password }}"
+    login_port: 4000
+    login_database: "admin"
+    replica_set: "{{ configsrv_replicaset }}"
+    validate: no
+    members:
+     - "localhost:4000"
+
+- pause:
+    seconds: 30
+
+- name: Add mongodb admin user to each shard 3.4+
+  mongodb_user:
+    login_host: localhost
+    login_port: "{{ item.port }}"
+    replica_set: "{{ item.rs }}"
+    database: admin
+    name: "{{ mongodb_admin_user }}"
+    password: "{{ mongodb_admin_password }}"
+    roles: ["root"]
+    state: present
+  register: mongo_admin_user
+  with_items:
+    - { "port": 3001, "rs": "{{ mongodb_replicaset1 }}" }
+    - { "port": 3004, "rs": "{{ mongodb_replicaset2 }}" }
+    - { "port": 4000, "rs": "{{ configsrv_replicaset }}" }
+  when: mongodb_version not in ["3.2", "4.0"] and test_mongo_auth == True
+
+- name: Add mongodb admin user to each shard 3.2 ^ 4.0
+  mongodb_user:
+    login_host: localhost
+    login_port: "{{ item.port }}"
+    replica_set: "{{ item.rs }}"
+    database: admin
+    name: "{{ mongodb_admin_user }}"
+    password: "{{ mongodb_admin_password }}"
+    roles: ["root"]
+    state: present
+  register: mongo_admin_user
+  with_items:
+    - { "port": 3001, "rs": "{{ mongodb_replicaset1 }}" }
+    - { "port": 3004, "rs": "{{ mongodb_replicaset2 }}" }
+  when: mongodb_version not in ["3.2", "4.0"] and test_mongo_auth == True
+
+# mongodb_user throws an error when creating a user on 3.2 (also on 4.0 with Ubuntu 18.04)
+# 'majority' is the only valid write concern when writing to config server replica sets
+- name: Copy create_user_root_3.2.js.j2 template to host
+  template:
+    src: create_user_root_3.2.js.j2
+    dest: /root/create_user_root_3.2.js
+  when: mongodb_version in ["3.2", "4.0"]
+
+- name: Create admin user on 3.2 and 4.0 config replset
+  shell: mongo admin --port 4000 /root/create_user_root_3.2.js
+  when: mongodb_version in ["3.2", "4.0"]
+
+- name: Murder all mongod processes
+  shell: pkill mongod && sleep 10 || sleep 10;
+
+- name: Execute mongod script to launch with auth shard1
+  script: files/bash/mongodb_replicaset.sh 3000 "{{ mongodb_replicaset1 }}" 3 1
+
+- name: Execute mongod script to launch with auth shard2
+  script: files/bash/mongodb_replicaset.sh 3003 "{{ mongodb_replicaset2 }}" 3 1
+
+- name: Launch cfg server with auth
+  command: mongod --configsvr --port 4000 --dbpath /home/tests/config --logpath /home/tests/config.log --smallfiles --replSet "{{ configsrv_replicaset }}" --fork --auth --keyFile /root/my.key
```

Comment:
Integration tests should use temporary directories whenever possible. Example:

https://github.com/ansible/ansible/blob/5ba7063f4fa2818681275b5237fe7e416d45f98b/test/integration/targets/old_style_modules_posix/meta/main.yml#L1-L2

Reply:
For this to work I'd need several temporary directories to be created and as far as I can tell that's not going to work here.

Reply:
Multiple temp directories can be created under the remote temp dir.

Reply:
OK, that might be possible but I'll have to manually remove them anyway several times as I create and teardown many instances of mongod. I'm not really sure what we gain out of it.

Reply:
setup_remote_tmp_dir now contains data dirs.